common/backoff/retry.go

```diff
@@ -92,29 +97,52 @@ func NewConcurrentRetrier(retryPolicy RetryPolicy) *ConcurrentRetrier {
 	return &ConcurrentRetrier{retrier: retrier}
 }
 
-// Retry function can be used to wrap any call with retry logic using the passed in policy
+// Retry function can be used to wrap any call with retry logic using the passed
+// in policy. A `nil` IsRetryable predicate will retry all errors. There is a
+// context-aware version of this function: RetryContext.
 func Retry(operation Operation, policy RetryPolicy, isRetryable IsRetryable) error {
+	ctxOp := func(context.Context) error { return operation() }
+	return RetryContext(context.Background(), ctxOp, policy, isRetryable)
+}
+
+// RetryContext is a context-aware version of Retry. Context
+// timeout/cancellation errors are never retried, regardless of IsRetryable.
+func RetryContext(
+	ctx context.Context,
+	operation OperationCtx,
+	policy RetryPolicy,
+	isRetryable IsRetryable,
+) error {
 	var err error
 	var next time.Duration
 
+	if isRetryable == nil {
+		isRetryable = func(error) bool { return true }
+	}
+
 	r := NewRetrier(policy, SystemClock)
-	for {
-		// operation completed successfully.  No need to retry.
-		if err = operation(); err == nil {
+	for ctx.Err() == nil {
+		//  completed successfully.  No need to retry.
+		if err = operation(ctx); err == nil {
 			return nil
 		}
 
 		if next = r.NextBackOff(); next == done {
 			return err
 		}
 
-		// Check if the error is retryable
-		if isRetryable != nil && !isRetryable(err) {
+		// Check if the error is retryable (err is known to be non-nil here)
+		if err == ctx.Err() || !isRetryable(err) {
 			return err
 		}
 
-		time.Sleep(next)
+		select {
+		case <-time.After(next):
```

Comment:
I think `time.After` has the potential to cause memory leaks.  Maybe not very likely in this case but still.  `time.After` timers are not garbage collected until they fire.  So if you break out of the loop due to `ctx.Done` then time.After timers will stay in memory until they fire.

Reply:
I was thinking that most retry processes never make it this far. But I can do it the other way if you think we're going to see a problem here.